### ISSUE TITLE
MetadataWriteFailed (warning) when selecting Restore read/write permissions

### ISSUE THREAD

**Issue by mediedata** [duplicati/duplicati#4028]
- [ x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info
- **Duplicati version**: 2.0.4.37_canary_2019-12-12
- **Operating system**: Win 10 pro
- **Backend**: B2

## Description

Got a Warning about null-value when restoring a backup, when the "Restore read/write permissions" option is checked. It happens only on one folder.

It doesn't happen in 2.0.4.23_beta_2019-07-14.

## Steps to reproduce

1. Restore a backup.
2. Select "Restore read/write permissions" and start restore

- **Actual result**:
Got a warning about null-value.
- **Expected result**:
No warning. The affected folder has the same read/write permissions as the other folders in the backup.

## Debug log

```
 "Warnings": [
    "2019-12-19 16.58.35 +01 
- [Warning-Duplicati.Library.Main.Operation.RestoreHandler-MetadataWriteFailed]: 
Failed to apply metadata to file: \"D:\\system\\Desktop\\restore test2\\_My Documents\\\", 
message: Value cannot be null.\r\nParameter name: sddlForm"
  ],
```

**Comment by warwickmm** [comment]
I can't reproduce this on Linux.  The parameter name `sddlForm` does not appear in the Duplicati source code, so I believe this is coming from somewhere else (perhaps in .NET somewhere).  A quick search seems to indicate that this may be Windows-specific (typically related to SQL server somehow).

**Comment by mediedata** [comment]
Yes it's definently a Windows-only thing. 

Digging a bit further, it looks like the issue is with reading/writing Windows' Security permissions on that folder.

The affected folder's permissions has an additional Deny-permission set for one user-account.
![Image1](https://user-images.githubusercontent.com/59032089/71290616-91581680-2370-11ea-99b9-d0b13bf4ef9b.jpg)

![Image2](https://user-images.githubusercontent.com/59032089/71290516-4fc76b80-2370-11ea-8902-2bed65e7096e.jpg)

It's most likely related to SetSecurityDescriptorSddlForm:
https://docs.microsoft.com/en-us/windows/win32/secauthz/security-descriptor-string-format
https://docs.microsoft.com/en-us/dotnet/api/system.security.accesscontrol.objectsecurity.setsecuritydescriptorsddlform?view=netframework-4.8

I'm trying to debug this before I change anything, but have an issue with "The breakpoint will not currently be hit no symbols have been loaded for this document" is Visual Studio.

**Comment by drwtsn32x** [comment]
Regarding the debugging issue in VS... make sure you have this environment variable set and restart VS:

AUTOUPDATER_Duplicati_SKIP_UPDATE=1

**Comment by mediedata** [comment]
Hi @drwtsn32x 

I'm moving the debug-issue out of this bug-report, since it's not related to normal usage.
To others: My debugging issue is located here: https://forum.duplicati.com/t/how-to-run-from-visual-studio-with-debugger/8700

**Comment by ts678** [comment]
Preliminary test here seems to show it's related to ACLs which aren't entirely inherited. You can see these or make these in Windows File Explorer in the Advanced option. These entries are encoded in the backup metadata file in a manner similar to the below. Three are for [well-known SIDs](https://support.microsoft.com/en-us/help/243330/well-known-security-identifiers-in-windows-operating-systems), and I obfuscate mine here:

```
{"win-ext:accessrules":"[{\"Rights\":1179785,\"ControlType\":0,\"SID\":\"S-1-1-0\",\"Inherited\":false,\"Inheritance\":0,\"Propagation\":0},{\"Rights\":2032127,\"ControlType\":0,\"SID\":\"S-1-5-18\",\"Inherited\":false,\"Inheritance\":0,\"Propagation\":0},{\"Rights\":2032127,\"ControlType\":0,\"SID\":\"S-1-5-32-544\",\"Inherited\":false,\"Inheritance\":0,\"Propagation\":0},{\"Rights\":2032127,\"ControlType\":0,\"SID\":XXX\",\"Inherited\":false,\"Inheritance\":0,\"Propagation\":0}]","CoreAttributes":"Archive","CoreLastWritetime":"636747209204933920","CoreCreatetime":"636747209204933920"}
```

Is DeserializeObject below going to try to write to variables that changed to public readonly due to [this](https://github.com/duplicati/duplicati/commit/2aa76881a51ce91762f7caabcc114434647532c2#diff-b77635aa9f9ba373452c972b0ca5b2df)? There was a problem somewhere else (I forget where) that was similar. I think it was also in a deserialize.

https://github.com/duplicati/duplicati/blob/1cc10b1caabb95f5e11c61904322d8085e80de3b/Duplicati/Library/Common/IO/SystemIOWindows.cs#L491-L493

If SID never gets set, the below might trigger the exception shown in some MS reference source below it.

https://github.com/duplicati/duplicati/blob/1cc10b1caabb95f5e11c61904322d8085e80de3b/Duplicati/Library/Common/IO/SystemIOWindows.cs#L86-L86

https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/mscorlib/system/security/principal/sid.cs#L394-L405

So there's an initial thought, but someone needs to try a test build (or one before the change), or debug. Testing 2.0.4.31 versus [v2.0.4.32-2.0.4.32_canary_2019-10-24](v2.0.4.32-2.0.4.32_canary_2019-10-24) on restore could be a no-build-needed test.

**Comment by ts678** [comment]
Previous fix for similar bug was [Remove readonly keyword from fields that are set by deserializer #3978](https://github.com/duplicati/duplicati/pull/3978).
[After last update duplicati doesn’t work anymore](https://forum.duplicati.com/t/after-last-update-duplicati-doesnt-work-anymore/8369/8?u=ts678) is on the perils of the automatic refactoring which was smart enough to make suggestions, but not smart enough to avoid wrong ones, thus breaking 2.0.4.32.

Tested by having 2.0.4.31 backup three empty files A.txt B.txt C.txt where A.txt was normal, inheriting its ACL, B.txt had Security --> Advanced --> Disable inheritance, and C.txt had Security --> Edit --> Add to give permission to Guest. 2.0.4.31 direct restore restored without an sddlForm warning but 2.0.4.32 did:

```
2019-12-23 07:44:27 -05 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-MetadataWriteFailed]: Failed to apply metadata to file: "C:\restore here\20191223\", message: The process cannot access the file 'C:\restore here\20191223' because it is being used by another process.
2019-12-23 07:44:27 -05 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-MetadataWriteFailed]: Failed to apply metadata to file: "C:\restore here\20191223\B.txt", message: Value cannot be null. Parameter name: sddlForm
2019-12-23 07:44:27 -05 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-MetadataWriteFailed]: Failed to apply metadata to file: "C:\restore here\20191223\C.txt", message: Value cannot be null. Parameter name: sddlForm
```

Both did the "being used by another process" on the folder, but I'm not sure if that's as big a deal as failing to do the Permissions restore that's off by default but which I checked. 2.0.4.32 left everything inherited, as if I hadn't checked the box. 2.0.4.31 added the non-inherited entries to inherited, which doesn't quite match the originals either. But even if it wasn't right before (I didn't test 2.0.4.5 beta), it probably is a little worse now. [How the System Uses ACLs](http://www.ntfs.com/ntfs-permissions-acl-use.htm) also explains how an ACL is an ordered list.

**Comment by drwtsn32x** [comment]
Looks like your hunch is correct, @ts678.

I did a VERY quick test where I removed the `readonly` from all these lines (didn't have time to narrow down which lines are actually the problem):

https://github.com/duplicati/duplicati/pull/3953/files#diff-b77635aa9f9ba373452c972b0ca5b2dfR61-R66

After that the restore worked better like the 2.0.4.31 version, but yes I also saw that the Duplicati doesn't properly restore the "inheritance broken" flag.  That's a different bug that has probably been around much longer.

Also I noticed that Duplicati would report the destination folder 'being used by another process' when I had the Explorer window open to that folder. When it was closed, that error was no longer generated.

**Comment by warwickmm** [comment]
Good work @ts678 and @drwtsn32x (and my fault for trusting the static analysis tools too much).  I think the lines that you both highlighted are the culprit.  The `FileSystemAccess` fields cannot be `readonly` since they are set by the deserializer.  I did a quick search for other instances where deserialization and `readonly` are incompatible and could not find any obvious ones.

@drwtsn32x, do you want to make the change since you have a Windows machine to test with?

**Comment by warwickmm** [comment]
I wonder if we can try keeping the `readonly` and use the `JsonProperty` attribute as described [here](https://stackoverflow.com/a/32888839) and [here](https://www.newtonsoft.com/json/help/html/SerializationGuide.htm).

In other words, I would prefer that we keep `readonly` if possible.  If not, that's ok too.

**Comment by drwtsn32x** [comment]
@warwickmm yes I can test and submit a PR. Will do later tonight or early tomorrow.

**Comment by warwickmm** [comment]
Thanks!

**Comment by drwtsn32x** [comment]
Using `[JsonProperty]` instead of removing `readonly` works. I will submit a PR with that being the fix.

Reminder - this does not address the 'being used by another process' message or the 'inheritance broken flag' not being restored properly.  We should open a new issue for the latter, if it doesn't exist already.

**Comment by warwickmm** [comment]
Fixed in pull request #4030.

**Comment by shinnlu** [comment]
i have the same issue when i choose "Restore read/write permissions"
Duplicati version: 2.0.6.3_beta_2021-06-17
Operating system: mac os 10.15.7

> 2021-07-02 00:18:24 +08 - [Warning-Duplicati.Library.Main.Operation.RestoreHandler-MetadataWriteFailed]: Failed to apply metadata to file: "/output/public/uploads/images/", message: Operation not permitted

i think the file/folder permission are not the problem, when i NOT choose "Restore read/write permissions", error disappeared.
